### ISSUE TITLE
help need in identify non-orthogonal crystal symmetry (e. g. honeycomb structure)

### ISSUE THREAD

**Issue by gjsdtckkk** [spglib/spglib#492]
Hi all, 
I am having issues with identifying a honeycomb structure with spglib in cpp. We know it should be P6/mmm, but in cpp, spglib (version 2.4.1 and 2.4.0) tells me that it is P2/m. However, in python, spglib (2.4.0) gives the correct prediction. Please find the included minimum codes for reproducing the issue.  Please help me on this issue, I need to use spglib in cpp.  Thanks you for your efforts.

**Comment by LecrisUT** [comment]
Please upload your source (codes.zip) to https://gist.github.com and share a link of that. Also edit your comment to remove the archive after that.

**Comment by gjsdtckkk** [comment]
Thanks, here is the cpp code to reproduce the issue
 https://gist.github.com/gjsdtckkk/6187e8a0d838c54ac4a79ad7c5708a1a

here is the python code
https://gist.github.com/gjsdtckkk/21c2054fe2cbf46746bd87e64315b33c

**Comment by lan496** [comment]
C-API assumes column-major basis vectors, and Python-API assumes row-major. So, can you try with a transposed matrix in the C-API case?
https://spglib.readthedocs.io/en/stable/python-interface.html#crystal-structure-cell

**Comment by LecrisUT** [comment]
Why is it even transposed? C is inheritly row-major, numpy default is row-major, the `position` is in row-major (need help checking `rotation`). The only place we should be transposing is in Fortran.

**Comment by lan496** [comment]
It is a historical reason in spglib. Literature in crystallography (like ITA) always uses row-major for basis vectors and C-API seems to adapt it. And later, other APIs seem to adapt column-major as programmer-friendly inputs.

**Comment by gjsdtckkk** [comment]
Yes, turn `lattice` to column-major works. My bad as I took it for granted when moving from python to c. Thanks to all of you!

**Comment by LecrisUT** [comment]
> It is a historical reason in spglib. Literature in crystallography (like ITA) always uses row-major for basis vectors and C-API seems to adapt it. And later, other APIs seem to adapt column-major as programmer-friendly inputs.

I needed a bit of time to think and articulate my thoughts on this. Basically this is a problem of interface definition: `lattice-matrix` vs `lattice-vectors`, i.e.:
```python
import numpy as np

lattice_vectors: list[np.array[float]] = [ np.array([a_x, a_y, a_z]), np.array([b_x, b_y, b_z]), np.array([c_x, c_y, c_z]) ]
lattice_matrix: np.array[float] = np.array(
    [ a_x, b_x, c_x
      a_y, b_y, c_y,
      a_z, b_z, c_z]).reshape((3, 3))
```

The issue of row-major vs column-major is actually irrelevant here because the `lattice_matrix` should have that mathematical form regardless. The issue is rather that we are allowing the interface of `list[list[float]]` which can construct either a `lattice_matrix` or a `lattice_vectors`, and we are opting to define the latter.

---

How to clarify this interface is rather tricky. One approach is to add explicit `kwargs` for either representation, but that conflicts with the tuple input that is used now. Maybe adding the `Lattice` datastructure (#489) would help with that.
```python
@dataclasses.dataclass
class Lattice:
    atoms: list[Atom]
    lattice: np.array[float]  = None
    _lattice_vectors: list[np.array[float] | np.ArrayLike[float]] = dataclasses.field(default=None, kw_only=True)
    _lattice_matrix: np.array[float] | np.ArrayLike[float] = dataclasses.field(default=None, kw_only=True)

    def __post_init__(self):
        self._check_inputs()  # make sure only one of `lattice`, `lattice_vectors`, `lattice_matrix` is defined
        if self._lattice_vectors:
            self.lattice = np.array(self._lattice_vectors).transpose()
        elif self._lattice_matrix:
            self.lattice = np.array(self._lattice_matrix)
        else:
            # some "smart" handling of ambiguous input
            if isinstance(self.lattice, list):
                # treat like lattice_vectors
                self.lattice = np.array(self.lattice).transpose()
            elif isinstance(self.lattice, np.array):
                # treat like lattice_matrix
                pass
            else:
                throw TypeError(f"Unsupported fromat for lattice [{type(lattice)}]")
```
